db/migrations/20240930100434_add_snap_uuid_to_templates.up.sql

```diff
@@ -0,0 +1,61 @@
+BEGIN;
+
+ALTER TABLE templates_repository_configurations ADD COLUMN IF NOT EXISTS snapshot_uuid UUID;
+
+-- migrate snapshots for use_latest templates
+UPDATE templates_repository_configurations trc
+SET snapshot_uuid = (
+    SELECT rc.last_snapshot_uuid
+    FROM repository_configurations rc
+    JOIN templates t ON t.uuid = trc.template_uuid
+    WHERE rc.uuid = trc.repository_configuration_uuid
+      AND t.use_latest = true
+      AND rc.last_snapshot_uuid IS NOT NULL
+)
+WHERE trc.template_uuid IN (
+    SELECT t.uuid
+    FROM templates t
+    WHERE t.use_latest = true
+);
+
+-- migrate snapshots for templates with a snapshot date
+UPDATE templates_repository_configurations trc
+SET snapshot_uuid = (
+    SELECT closest_snapshots.uuid
+    FROM (
+         (SELECT s.uuid, s.created_at
+          FROM snapshots s
+          JOIN templates t ON t.uuid = trc.template_uuid
+          WHERE s.repository_configuration_uuid = trc.repository_configuration_uuid
+            AND t.use_latest = false
+            AND s.created_at <= t.date
+          ORDER BY s.created_at DESC
+          LIMIT 1)
+
+         UNION
+
+         (SELECT s.uuid, s.created_at
+          FROM snapshots s
+          JOIN templates t ON t.uuid = trc.template_uuid
+          WHERE s.repository_configuration_uuid = trc.repository_configuration_uuid
+            AND t.use_latest = false
+            AND s.created_at > t.date
+          ORDER BY s.created_at ASC
+          LIMIT 1)
+     ) AS closest_snapshots
+    ORDER BY closest_snapshots.created_at ASC
+    LIMIT 1
+)
+WHERE trc.template_uuid IN (
+    SELECT t.uuid
+    FROM templates t
+    WHERE t.use_latest = false
+);
+
+ALTER TABLE templates_repository_configurations
+DROP CONSTRAINT IF EXISTS fk_templates_repository_configurations_snapshots,
+ADD CONSTRAINT fk_templates_repository_configurations_snapshots
+FOREIGN KEY (snapshot_uuid) REFERENCES snapshots(uuid)
+ON DELETE RESTRICT;
+
+COMMIT;
```

Comment:
do we want to make  the `snapshot_uuid` column not null?

Reply:
i think so, do you think we should do that here or in a following PR? 

to do that we'd need to make sure all repos have snapshots before they're added to a template (which i've added in here). we'd also need to set not null on the column and the model, and since there would be a few seconds where the column would be null until the update-template-content task runs and adds in the snapshot uuids, creating a template would fail. so we'd probably need to add a method to insert the snapshots before the template is created. (and there might be some additional steps that i'm missing)

Reply:
pushed up these changes!